### ISSUE TITLE
Improved tools for 3D annotation

### ISSUE THREAD

**Issue by jni** [napari/napari#5955]
- [x] ~~Add, remove, edit labels in 3D~~
- [ ] Slice interpolation (nD) (plugins?) (see jni/zarpaint#32 and https://github.com/brisvag/napari-label-interpolator)
- [ ] 3D Labels editing performance for large volumes
- [ ] Add labels, points, or polygons on a clipping or slicing plane
- [ ] Add and edit 3D bounding boxes
- [ ] Manipulate points and shapes in 3D
- [ ] Add points or labels based on intensity of an underlying image or position of polygons or surfaces in other layers

**Comment by jni** [comment]
@GenevieveBuckley I incorporated your comments in the draft into two extra items above. I actually strongly disagree that slice interpolation should be built-in rather than in plugins: there are umpteen methods for slice interpolation and it's not particularly clear which method to implement as a built-in. However, because our slicing model is so bad, right now it's pretty clunky to do slice interpolation in a plugin, so I definitely understand the desire to fold it into main. But I think instead that we should focus on improving our slice models (#5949) and expand  our plugin infra/contribution types (#5874) until slice interpolation in both labels and shapes layers is really nice and easy to implement in plugins.

At any rate, we can discuss, I'm definitely not feeling dogmatic about it. But I would very much like to tighten the scope of napari core so we can get some of these long-standing issues sorted, before adding (potentially) complicated algorithms to core.

Regarding performance, have you tried 3D editing recently? Contributions like #5723 / #5732 might have made a significant dent in the user experience. And of course there's #3308 in the pipeline... I hope I can finish that up soon... 😩

**Comment by GenevieveBuckley** [comment]
> Regarding performance, have you tried 3D editing recently? Contributions like #5723 / #5732 might have made a significant dent in the user experience. And of course there's #3308 in the pipeline... I hope I can finish that up soon... 😩

I've been trying to avoid 3D label editing, but in the last fortnight I was editing labels for a large tiled 2D image. It was very, very slow. Almost unusable, given the lag time between moving the paint/erase brush and actually seeing the response to the pixels on screen. I suspect that the plugin might be able to try something more efficient than working with the default new label layer, which might help as well.

EDIT: I've just looked at what napari version I was using at the time, and it's 0.4.17. Looks like there hasn't been a new release for the last seven months, so I think it's safe to say neither of the PRs you linked were in effect. https://github.com/napari/napari/pull/5732 might speed things up a bit, I'll have to try it (but https://github.com/napari/napari/pull/5723 seems irrelevant to this discussion).

**Comment by jni** [comment]
> Looks like there hasn't been a new release for the last seven months, so I think it's safe to say neither of the PRs you linked were in effect.

Correct. Let us know if you have any issues running from main. Also Grzegorz is gearing up to release 0.4.18 which includes those PRs. Should be out in 1-2 weeks. (Has to happen cos I want to use it at SciPy 😂)

**Comment by bauerdavid** [comment]
I have a plugin which implements 3D bounding boxes as a custom layer: [napari-bbox](https://www.napari-hub.org/plugins/napari-bbox). There are lots of hacky solutions involved, but the basics are there (2D-3D display, displaying bounding boxes only on slices relevant, editing/creating in 2D etc.). I would be glad to contribute in making it a built-in layer, however, I have only a limited time to work on it. But if someone could join, we can make it work 🙂.

The code by the way is an edited version of the shapes layer (from a previous version of napari, `0.4.15` if I'm correct).

**Comment by jni** [comment]
Thanks for chiming in @bauerdavid! What do you think about incorporating the boxes directly into the Shapes layer? They seem to be fundamentally related...?

**Comment by brisvag** [comment]
@bauerdavid I thnk we already talked about this elsewhere before, but in my opinion this should be an overlay. I know that (differently to the already implemented *layer* bounding boxes) your plugin also provides the ability to give arbitrary bounding boxes; the main question is: do we consider this as "data" per se, or does it only make sense as a "thing that highlights specific regions in another layer"? If it's the second, overlays are the way to go IMO. Depending on the use case, it might actually be time to hook up and use the [interaction box overlay](https://github.com/napari/napari/blob/main/napari/components/overlays/interaction_box.py).

Alternatively (if the bounding boxes don't need specific interactivity or visualisation tricks), they could also just be helper functions on the shapes layer that can take "something" (another layer, a selection of points, etc) and generate a n-D bounding box around it.

**Comment by bauerdavid** [comment]
Hey @jni,
In my opinion, there are some reasons which make it necessary to separate shapes from bounding boxes:
1. Shapes are "fundamentally 2D". If a Shapes layer is 3D, it still means that you are placing 2D objects (which lie on planes) in 3D.
2. (This is connected to the previous point). Actually you can modify the points of shapes to not lie on a plane, but still in a shape every anchor point is connected with (at most) 2 other points, while in a 3D bounding box every point is connected to 3 other (As we increase the number of dimensions, the number of neighbor points increases, but maybe annotating 4+dimensional images, excluding time and channels, might not be a thing 😄.)
3. Rendering 3D bounding boxes needs extra care (e.g. in 2D display bounding boxes only if the current slice is inside the span of the bounding box, etc.).
4. For creating 2D bounding boxes, you could use the Shapes layer. However, in that case editing each point separately (thus turning a rectangle into a general shape) is unnecessary and allows some user error to occur. This is also true rotating bounding boxes (although I can imagine, that in some cases this might be needed).
I think the above points make bounding boxes incompatible with the Shapes layer. What do you think?

Hey @brisvag,
I agree @jni, there are some cases when you don't have any underlying data (like labels or shapes), so I do think that bounding boxes can be considered data in themselves. However, I also like the idea of having a bounding box displayed for some other data ,e.g. highlight the currently selected label in a Labels layer. Actually I use it that way as well, but that would be really nice if it was a property for layers.

**Comment by brisvag** [comment]
I totally forgot that our shapes don't allow for 3D objects... well, ignore that part of my comment xD

>  highlight the currently selected label in a Labels layer

Yeah this is where overlays would be awesome I think. But I agree that many other cases fall in the "data" world. Is there a reason why we call them "bounding boxes" (which in my mind sounds like a *property* of some data) versus for instance ROIs, which sounds more like what you're talking about? Or am I mixing up concepts?

**Comment by bauerdavid** [comment]
I agree, "ROI" might be a better phrase for this concept. Strangely enough, when I saw discussions for 3D ROI's, I usually saw (and used) the term "bounding box"... 🤔

**Comment by jni** [comment]
> Shapes are "fundamentally 2D". If a Shapes layer is 3D, it still means that you are placing 2D objects (which lie on planes) in 3D.

> Actually you can modify the points of shapes to not lie on a plane, but still in a shape every anchor point is connected with (at most) 2 other points, while in a 3D bounding box every point is connected to 3 others

that's true now but not necessarily true in the future. They are 2D because 2D was easier to implement, not because we thought they should be 2D forever. 

> Rendering 3D bounding boxes needs extra care (e.g. in 2D display bounding boxes only if the current slice is inside the span of the bounding box, etc.).

Again, I think "fancy 3D shape" can define all of the attributes you want that "boring 2D shape" does not. It's just an accident of history. The question is, do we want to have one layer for each dimensionality of shape (in which case we should move "paths" out of Shapes into a Shapes1D layer???), or do we want to have a "vector shapes" layer that encompasses all the dimensions — and rendering options and editing tools appropriate to each shape?

> However, in that case editing each point separately (thus turning a rectangle into a general shape) is unnecessary and allows some user error to occur

Even things like moving the vertex of a rectangle I think are supported by an accident of history, not because we thought it was particularly useful to turn rectangles into general polygons. I agree that it's annoying, it's bitten me a couple of times!

See above @bauerdavid, I don't actually think of your points as "fundamental limitations of the Shapes layer" but "accidents of history unfortunately on the Shapes layer", and I think adding 3D bounding boxes would *help* clean up the design choices of the Shapes layer.

**Comment by bauerdavid** [comment]
> See above @bauerdavid, I don't actually think of your points as "fundamental limitations of the Shapes layer" but "accidents of history unfortunately on the Shapes layer", and I think adding 3D bounding boxes would help clean up the design choices of the Shapes layer.

Yeah, you're right, the points I wrote are more like implementation issues. What you wrote made me rethink the question from a functionality point of view.
Bounding boxes (or ROI's) can only be rectangles or cuboids (or hypercuboids for higher dimensions, but let's not go there right now 😄). They are used to create boxes, which contain some objects. On the other hand, with shapes you can draw the boundaries of the said object precisely. (Duh... 😄)
Now, I think the following questions should be thinked through (from the least important to the most relevant):

1.
    a) Would users of napari want to draw 3D shapes manually?
    b) What uses would we see for that in image processing?
2. IMO the Surface layer in napari can be considered the (true) 3D version of Shapes layer. Should these be merged?
    > **Note** At first sight the existence of the Surface layer might answer 1/b). I never used it, but if I'm correct, it's basically used for visualization, not for 3D editing or annotation. But correct me if I'm wrong.
3. It's important that 3D bounding boxes can be edited easily in 2D, even if we are not on the slice where the bounding box corners are. (Imagine that you want to resize the box to be the minimal bounding box of an object. You should be able to drag the sides of the box in the middle of it.) How would you generalize this for any 3D shape? How would you edit a general 3D shape in 2D? Or only boxes should be edited in 2D, anything else only in 3D?
4. If ROI's are part of the Shapes layer, should editing limitations be applied to bounding boxes (no object-wise rotation, no point-wise editing), or left as it is now (see rectangles)?
5. (Follow-up from 3. and 4.) Why should ROI's be part of the Shapes layer, if they need special care?

**Comment by brisvag** [comment]
> 1.
>    a) Would users of napari want to draw 3D shapes manually?
>    b) What uses would we see for that in image processing?

3D paths, for sure. Things like spheres/ellipses and cubes also likely. Not sure what those would actually be in terms of data, but if I could use them I would :P

> 2. IMO the Surface layer in napari can be considered the (true) 3D version of Shapes layer. Should these be merged?
>    > **Note** At first sight the existence of the Surface layer might answer 1/b). I never used it, but if I'm correct, it's basically used for visualization, not for 3D editing or annotation. But correct me if I'm wrong.

I disagree; shapes are mathematical descriptions of objects (more "abstract", which allows for different kinds of analysis and uses). Surfaces are also a lot harder to generate and work with in terms of data. There's also several things that surfaces simply couldn't represent (such as paths).

> 3. It's important that 3D bounding boxes can be edited easily in 2D, even if we are not on the slice where the bounding box corners are. (Imagine that you want to resize the box to be the minimal bounding box of an object. You should be able to drag the sides of the box in the middle of it.) How would you generalize this for any 3D shape? How would you edit a general 3D shape in 2D? Or only boxes should be edited in 2D, anything else only in 3D?

Definitely agree! If you see how the `TransformBox` is implemented (set a layer's mode to `transform` that's exactly how it works. As for general 3D shapes, I don't think we can generalize like that. Something like a path, an ellips(oid), or a box will never overlap perfectly in terms of data and how they can be manipulated. Changing the radii of an ellipse makes sense, not for a path. I think we need to embrace the variability of this layer, rather than force a unified behaviour. If it gets *too* strange, we can always split it and create sublayers...

> 4. If ROI's are part of the Shapes layer, should editing limitations be applied to bounding boxes (no object-wise rotation, no point-wise editing), or left as it is now (see rectangles)?
> 5. (Follow-up from 3. and 4.) Why should ROI's be part of the Shapes layer, if they need special care?

If we think of them as "n-dimensional parallelepipeds", they are just another type of shape, and as I was mentioning above, each shape has its own limitations and behaviours already.

**Comment by bauerdavid** [comment]
Okay, I think I start to understand the concept 😄. So basically there would be a layer, which can handle regular objects (spheres, ellipsoids, "n-dimensional parallelepipeds") both in 2D and 3D; and irregular curves, in which every "anchor point" (I don't know if this is a valid word here) has exactly two other points connected, except for open curves, where the two end points have a single connection. Both ellipsoids and parallelepipeds could be defined by their bounding boxes. I suppose there would be no irregular 3D object (by that I mean objects which have a volume, unlike 3D curves). Does this seem correct?

> I disagree; shapes are mathematical descriptions of objects (more "abstract", which allows for different kinds of analysis and uses). Surfaces are also a lot harder to generate and work with in terms of data. There's also several things that surfaces simply couldn't represent (such as paths).

Yeah, this makes sense. My confusion came from the assumption that, for example an ellipsoid would be described by a point cloud in a 3D shapes layer. But I guess they can be defined similarly to the currently available ellipse in Shapes layer (by a bounding box).

> I think we need to embrace the variability of this layer, rather than force a unified behavior. If it gets _too_ strange, we can always split it and create sublayers...

I can imagine a Shapes layer with custom behavior for each kind of shape. However, I'm not sure if this is the best option. Again, I can be wrong, but maybe from the UX point of view, it would be clearer to separate different kinds of shapes into separate (sub)layers. Would a user want to annotate ellipsoids, boxes and free-hand curves in the same layer? As a plugin developer, I would be happier if I am sure that the users can only create the types of objects in a layer I want them to create (e.g. boxes). And also, usually if someone annotates ROI's, they want the boxes to have the same orientation as the image grid (no 42.1783° rotation). But of course, I can imagine a checkbox/API for the layer to disable rotation.

I think this is more of a "religious" problem, I can see either option - unified shapes layer vs. separate layer for each type -working, but I'm more for the latter option.
